code/src/gfx.c

```diff
@@ -377,36 +377,46 @@ static void Gfx_DrawSpoilerItemGroups(void) {
     if (gSettingsContext.shuffleOverworldEntrances || gSettingsContext.shuffleDungeonEntrances) {
-        u16 itemCount = ENTRANCE_OVERRIDES_MAX_COUNT;
+        u16 itemCount = gEntranceTrackingData.EntrancePairsCount;
         u16 firstItem = entranceScroll + 1;
         u16 lastItem = entranceScroll + MAX_ITEM_LINES;
         if (lastItem > itemCount) { lastItem = itemCount; }
         Draw_DrawFormattedString(10, 10, COLOR_TITLE, "Randomized Entrances (%d - %d) / %d",
             firstItem, lastItem, itemCount);
 
-        // TODO random entrances visited %
+        // TODO random entrances visited percent?
         // Draw_DrawFormattedString(SCREEN_BOT_WIDTH - 10 - (SPACING_X * 6), 10, itemPercent == 100 ? COLOR_GREEN : COLOR_WHITE, "%5.1f%%", itemPercent);
 
         u16 listTopY = 26;
         for (u32 item = 0; item < MAX_ITEM_LINES; ++item) {
             u32 locIndex = item + entranceScroll;
             if (locIndex >= itemCount) { break; }
 
-            EntranceOverride entrOverride = rEntranceOverrides[locIndex];
-            if (entrOverride.index == 0 && entrOverride.override == 0) {
-                continue;
-            }
-
             u32 locPosY = listTopY + ((SPACING_SMALL_Y + 1) * item * 2);
             u32 itemPosY = locPosY + SPACING_SMALL_Y;
-            bool isDiscovered = SaveFile_GetIsEntranceDiscovered(entrOverride.index);
+
+            EntranceTrackingPair pair = gEntranceTrackingData.EntrancePairs[locIndex];
+            bool isDiscovered = SaveFile_GetIsEntranceDiscovered(pair.StartIndex) || SaveFile_GetIsEntranceDiscovered(pair.ReturnIndex);
+
+            if (!isDiscovered) {
+                // If the pair included one of the hyrule field <-> zora's river entrances,
+                // the randomizer will have also overriden the water-based entrances, so check those too
+                if ((pair.StartIndex == 0x00EA || pair.ReturnIndex == 0x00EA) && SaveFile_GetIsEntranceDiscovered(0x01D9)) {
+                    isDiscovered = true;
+                } else if ((pair.StartIndex == 0x0181 || pair.ReturnIndex == 0x0181) && SaveFile_GetIsEntranceDiscovered(0x0311)) {
+                    isDiscovered = true;
+                }
+            }
+
             u32 color = isDiscovered ? COLOR_GREEN : COLOR_WHITE;
-            Draw_DrawFormattedString_Small(10, locPosY, color,
-                "0x%04X -> 0x%04X, BW: 0x%04X", entrOverride.index, entrOverride.override, entrOverride.blueWarp);
+            const char* unknown = "???";
+            const char* startName = pair.StartStrOffset != ENTRANCE_INVALID_STRING_OFFSET ? &gEntranceTrackingData.StringData[pair.StartStrOffset] : "START STRING NOT FOUND";
+            const char* returnName = pair.ReturnStrOffset != ENTRANCE_INVALID_STRING_OFFSET ? &gEntranceTrackingData.StringData[pair.ReturnStrOffset] : "RETURN STRING NOT FOUND";
+            Draw_DrawFormattedString_Small(10, locPosY, color, "%s ->", isDiscovered ? startName : unknown);

```

Comment:
```suggestion
            Draw_DrawFormattedString_Small(10, locPosY, color, "%s <->", isDiscovered ? startName : unknown);
```
Since entrances are always coupled and the list doesn't have an entry for the entrance on the 2nd row.